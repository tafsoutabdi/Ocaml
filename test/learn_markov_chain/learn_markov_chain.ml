open MarkovChain.Learner
open MarkovChain.RandomWalk
open Definitions.MarkovChain
open Utils.Std
open Utils.MarkovChain

let () = Random.self_init ()

let token_of_arc src dst = match (src, dst) with
| (0, 1) -> 'a'
| (1, 2) -> 'g'
| (1, 3) -> 'b'
| (3, 4) -> 'e'
| (3, 5) -> 'c'
| (3, 2) -> 'd'
| (4, 1) -> 'f'
| _ -> invalid_arg "token_of_arc"

let () = 
  let samples = [[0; 1]; [1; 3; 4; 1; 2]; [3; 4]; [3; 5]; [3; 5]; [3; 5]] in
  let mc_estimated = learn_markov_chain ~token_of_arc ~max_state_id: 5 ~walks: samples in
  Format.printf "learn_markov_chain token_of_arc 5 %s =@.%s@.@."
    (string_of_list (string_of_list string_of_int) samples)
    (string_of_markov_chain mc_estimated)

let runtest var_name (MarkovChain mc) string_of_token nb_samples = 
  Format.printf "apprentissage à partir de marches aléatoires aléatoires sur %s`@." var_name;
  let proba_of_arc (MarkovChain mc) src dst = 
    try
      let total_weight = 
        mc.(src) |> List.map (fun {weight; _} -> weight) |> List.fold_left (+) 0 in
      let weight = (List.find (fun {dest; _} -> dest=dst) mc.(src)).weight in
      percent_of_frac weight total_weight
    with
      _ -> 0 in
  let max_state_id = Array.length mc in
  Format.printf "echantillonage...@.";
  let samples = 
    let mc2 = mc |> Array.map (List.map (fun (edge) -> {edge with token=edge.dest})) in
    List.init nb_samples (fun _ -> 0 ::
      random_walk ~length:(Random.int 20) (MarkovChain mc2)) in
  let token_of_arc src dst = 
    (List.find (fun {dest; _} -> dest=dst) mc.(src)).token in
  Format.printf "apprentissage...@.";
  let mc_estimated = learn_markov_chain ~token_of_arc ~max_state_id ~walks: samples in
  Format.printf "résultat@.";
  for i=0 to max_state_id - 1 do
    mc.(i) |> List.iter (fun {token;dest; _} -> 
      Format.printf "arc %d-%s->%d : proba estimée %d%%, proba réelle %d%%@."
       i (string_of_token token) dest (proba_of_arc mc_estimated i dest) (proba_of_arc (MarkovChain mc) i dest)
    )
  done;
  Format.printf "@."

let () = runtest "mc_smaller_example" mc_smaller_example (fun s -> s) 1_000_000
  
let () = runtest "mc_example" mc_example Char.escaped 1_000_000