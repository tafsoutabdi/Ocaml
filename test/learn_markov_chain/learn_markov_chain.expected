learn_markov_chain token_of_arc 5 [[0; 1]; [1; 3; 4; 1; 2]; [3; 4]; [3; 5]; [3; 5]; [3; 5]] =

[|[{token='a'; weight=1; dest=1}];
[{token='b'; weight=1; dest=3}; {token='g'; weight=1; dest=2}];
[];
[{token='c'; weight=3; dest=5}; {token='e'; weight=2; dest=4}];
[{token='f'; weight=1; dest=1}];
[]|]


apprentissage à partir de marches aléatoires aléatoires sur mc_smaller_example`
echantillonage...
apprentissage...
résultat
arc 0-a->0 : proba estimée 50%, proba réelle 50%
arc 0-b->1 : proba estimée 50%, proba réelle 50%
arc 1-c->1 : proba estimée 67%, proba réelle 67%
arc 1-d->2 : proba estimée 33%, proba réelle 33%

apprentissage à partir de marches aléatoires aléatoires sur mc_example`
echantillonage...
apprentissage...
résultat
arc 0-a->1 : proba estimée 100%, proba réelle 100%
arc 1-b->3 : proba estimée 50%, proba réelle 50%
arc 1-g->2 : proba estimée 50%, proba réelle 50%
arc 3-c->5 : proba estimée 50%, proba réelle 50%
arc 3-d->2 : proba estimée 17%, proba réelle 17%
arc 3-e->4 : proba estimée 33%, proba réelle 33%
arc 4-f->1 : proba estimée 100%, proba réelle 100%

